src/components/common/TokenImg.tsx

```diff
@@ -0,0 +1,76 @@
+import React from 'react'
+import styled from 'styled-components'
+
+import { getImageUrl, safeTokenName } from 'utils'
+import unknownTokenImg from 'assets/img/unknown-token.png'
+
+const Wrapper = styled.img<WrapperProps>`
+  width: 2.8rem;
+  height: 2.8rem;
+  border-radius: 3.6rem;
+  object-fit: contain;
+  background-color: white;
+  padding: 2px;
+  opacity: ${(props): number => (props.faded ? 0.4 : 1)};
+`
+
+function _loadFallbackTokenImage(event: React.SyntheticEvent<HTMLImageElement>): void {
+  const image = event.currentTarget
+  image.src = unknownTokenImg
+}
+
+export interface Props {
+  symbol?: string
+  name?: string
+  address: string
+  addressMainnet?: string
+  faded?: boolean
+}
+
+export interface WrapperProps {
+  faded?: boolean
+}
+
+const tokensIconsRequire =
+  process.env.NODE_ENV === 'test'
+    ? // mock for jest
+      Object.assign(() => '', {
+        keys: () => [],
+        resolve: () => '',
+        id: '',
+      })
+    : require.context('assets/img/tokens', false)
+const tokensIconsFiles = tokensIconsRequire.keys()
+
+export const TokenImg: React.FC<Props> = (props) => {
+  const { address, addressMainnet, symbol, name } = props
+
+  // TODO: Probably a good idea to allow to see in etherscan (probably wrapping this in BlockExplorerLink component)
+
+  // TODO: Simplify safeTokenName signature, it doesn't need the addressMainnet or id!
+  // return walletIconFile && tokensIconsRequire<{ default: string }>(walletIconFile).default
+
+  const iconFile = tokensIconsFiles.find(
+    (file) => file.includes(address) || (addressMainnet && file.includes(addressMainnet)),
+  )
```

Comment:
Since you can rely on token addresses being a static string (unlike wallet names that can be anything a wallet returns), you can simplify the lookup to an object, Map or even a Set

```ts
// Set<'0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d'>
const tokensIconsFilesSet = tokensIconsRequire.keys().map(filename => filename.replace(/^\.\//, '')..replace(/\.\w+$/, ''))

// then 
const iconFile = tokensIconsFilesSet.has(address) && './' + address + '.png'
// handle addressMap
```

Now that I think about it better to have an object,so we don't rely on a fixed file extension
```ts
// address => file mapping, made from tokensIconsRequire.keys()
const tokensIconsFiles = {
  '0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d': './0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d.png'
}

// then
const iconFile = tokensIconsFiles[address] || tokensIconsFiles[addressMainnet]
```

Whatever you choose,it makes O(n) lookup with find on each render into O(1)

Reply:
done